Add codemod for border style compatibility

RobinMalfait opened the pull request:
This PR adds a codemod that ensures that the border styles from Tailwind CSS v3 work as expected once your project is migrated to Tailwind CSS v4.

In Tailwind CSS v3, the default border color is `colors.gray.200` and in Tailwind CSS v4 the default border color is `currentColor`.

Similarly in Tailwind CSS v3, DOM elements such as `input`, `select`, and `textarea` have a border width of `0px`, in Tailwind CSS v4, we don't change the border width of these elements and keep them as `1px`.

If your project happens to already use the same value for the default border color (`currentColor`) as we use in Tailwind CSS v4, then nothing happens. But this is very unlikely, so we will make sure that we honor your `borderColor.DEFAULT` value.

If you didn't change the default values in your `tailwind.config.js`, then we will inject compatibility CSS using the default Tailwind CSS v3 values to ensure the default color and width are applied correctly.

RobinMalfait commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/tailwindlabs/tailwindcss/14746?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#14746** <a href="https://app.graphite.dev/github/pr/tailwindlabs/tailwindcss/14746?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* **#14745** <a href="https://app.graphite.dev/github/pr/tailwindlabs/tailwindcss/14745?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `next`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @RobinMalfait and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

RobinMalfait commented on packages/@tailwindcss-upgrade/src/codemods/migrate-border-compatibility.ts:
```diff
@@ -0,0 +1,213 @@
+
+const css = dedent
+const COMPATIBILITY_CSS = css`
+  /**
+   * The default border styles in Tailwind CSS v4 changed compared to 
+   * Tailwind CSS v3. To make sure your project looks the same, we've added
+   * these compatibility styles to your project.
+   *
+   * It is recommended to migrate your project, and delete this layer once you
+   * are done.
+   *
+   * - All elements with a 'border' class will default to 'currentColor' in
+   *   Tailwind CSS v4, instead of 'border-gray-200' in Tailwind CSS v3.
+   * - All inputs, selects, and textareas will have a 'border-width: 1px' in
+   *   Tailwind CSS v4, instead of 'border-width: 0' in Tailwind CSS v3.
+   */
```
Can probably simplify this, but wanted to include _why_ we are doing this and how you could/should proceed.

RobinMalfait commented on packages/@tailwindcss-upgrade/src/migrate.ts:
```diff
@@ -36,12 +37,13 @@ export async function migrateContents(
     .use(migrateMissingLayers())
     .use(migrateTailwindDirectives(options))
     .use(migrateConfig(stylesheet, options))
+    .use(migrateBorderCompatibility(options))
```
Requires the tailwind directives to `@import` migrations

RobinMalfait commented on packages/@tailwindcss-upgrade/src/migrate.ts:
```diff
@@ -36,12 +37,13 @@ export async function migrateContents(
     .use(migrateTailwindDirectives(options))
     .use(migrateConfig(stylesheet, options))
+    .use(migrateBorderCompatibility(options))
+    .use(migrateThemeToVar(options))
```
Using `theme(…)` inside of the compatibility CSS, so needed to move this after the `migrateBorderCompatibility` migration.

RobinMalfait commented on packages/@tailwindcss-upgrade/src/codemods/migrate-border-compatibility.ts:
```diff
@@ -0,0 +1,200 @@
+   * - All inputs, selects, and textareas will have a 'border-width: 1px' in
+   *   Tailwind CSS v4, instead of 'border-width: 0' in Tailwind CSS v3.
+   */
+  @layer base {
```
In my testing the `@layer base` isn't strictly required (once the previous PR where we updated `preflight.css` is in place). But I think this is still nicer because:

1. It is about the reset that already lives in the `base` layer
2. It groups the compatibility CSS together, which is nice if you look at the CSS files we've updated

RobinMalfait commented on integrations/upgrade/js-config.test.ts:
```diff
@@ -468,3 +642,218 @@ test(
+          ::before,
+          ::backdrop,
+          ::file-selector-button {
+            border-color: var(--border-color, var(--color-gray-200, currentColor));
```
This now contains `var(--border-color)` because it existed in your own tailwind.config.js file. We could also always generate this instead of just `var(--color-gray-200, currentColor)`.

philipp-spiess replied:
@RobinMalfait IIRC in the call yesterday we discussed maybe inlining the value of `--border-color` and omitting it from the created CSS (since it doesn't do anything for the core utilities anyways). WDYT about that approach?

RobinMalfait replied:
Yeah not a bad idea. I started with resolving the value using `theme(…)` and since it generated the variable and used it looked nice because you can still update the default.

However, the moment you get rid of the backwards compatibility CSS, then the variable is useless and vice versa.

Let me inline it instead.

RobinMalfait commented on integrations/upgrade/js-config.test.ts:
```diff
@@ -468,3 +642,218 @@ test(
+          textarea {
+            border-width: 0;
+          }
+        }
```
Notice that the default styles for the `border-color` are not here because they resolve to `currentColor` which is also the default in v4. Therefore if we would've added it, it would be additional CSS doing the same thing.

philipp-spiess commented on packages/@tailwindcss-upgrade/src/codemods/migrate-border-compatibility.test.ts:
```diff
@@ -0,0 +1,235 @@
+
+const css = dedent
+
+async function migrate(input: string) {
```
Let's add a test here as well with a `Config` that does define `borderColor.DEFAULT` and one where said value uses a `theme()` function. I saw code related to it in your implementation 🙂

RobinMalfait replied:
In this file it's a bit tricky since we resolve user configs, so I added dedicated tests for all of these in the integration tests:
https://github.com/tailwindlabs/tailwindcss/blob/e6c13b968986f0469fe7d67aa78c4d2c213978be/integrations/upgrade/js-config.test.ts#L646

The backwards compatibility CSS uses the `theme(…)` function, which gets replaced which you can also see in that file above.

I can of course add some tests to this file as well where I mock the user config. But it was easier to write integration tests for.

philipp-spiess replied:
oh I skipped over this file since there were so many diffs because of the new comments 🙈 Thanks!

RobinMalfait replied:
Thank you GitHub for hiding the diff 😂

philipp-spiess reviewed (commented):
LGTM, but we may need to tweak the messaging a bit. I provided a suggestion inline

philipp-spiess commented on integrations/upgrade/js-config.test.ts:
```diff
@@ -468,3 +642,218 @@ test(
+            border-width: 0;
+          }
+        }
+
+        @theme {
+          --border-color: currentColor;
+        }
```
might not need that in this case either

RobinMalfait replied:
This is a side effect of another codemod where we convert the js config to a CSS-based `@theme` config. 

We have another card in Basecamp where we should purge these kinds of unnecessary values so I feel like handling this where we remove it should not be part of this codemod. 🤔

philipp-spiess commented on packages/@tailwindcss-upgrade/src/index.test.ts:
```diff
@@ -149,6 +184,39 @@ it('should migrate a stylesheet (with imports)', async () => {
+      ::file-selector-button {
+        border-color: var(--color-gray-200, currentColor);
+      }
+    }
+    /*
```
```suggestion
    }

    /*
```

Seems like something screws up the empty line between the two blocks 😢

RobinMalfait replied:
Figured it out. In this case the test is performing migrations and once done formats the nodes on a stringified version. The actual integration tests don't have this issue.

philipp-spiess replied:
If it works in the integration tests this is fine!

philipp-spiess commented on packages/@tailwindcss-upgrade/src/index.test.ts:
```diff
@@ -175,6 +243,38 @@ it('should migrate a stylesheet (with preceding rules that should be wrapped in
     /**! My license comment */
     @import 'tailwindcss';
     @config './tailwind.config.js';
+    /*
```
Unsure why there is no empty line above this one here?

RobinMalfait replied:
Fun part is that the space between the `@import` and `@config` is also gone while it should be there. 

So I feel like whitespace handling might need a dedicated PR to get it all right. But I'll take a look at the other cases to see if it's easily fixable or not.

Bigger problem is that in isolation it's correct, but the accumulation of changes is not.

RobinMalfait replied:
Figured it out. In this case the test is performing migrations and once done formats the nodes on a stringified version. The actual integration tests don't have this issue.

philipp-spiess commented:
### Merge activity

* **Oct 22, 12:07 PM EDT**: The merge label 'merge-queue' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=tailwindlabs&repo=tailwindcss) once it meets the requirements.

RobinMalfait commented on CHANGELOG.md:
```diff
@@ -11,6 +11,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 - _Upgrade (experimental)_: Migrate `plugins` with options to CSS ([#14700](https://github.com/tailwindlabs/tailwindcss/pull/14700))
 - _Upgrade (experimental)_: Allow JS configuration files with `corePlugins` options to be migrated to CSS ([#14742](https://github.com/tailwindlabs/tailwindcss/pull/14742))
+- _Upgrade (experimental)_: Add codemod for border style compatibility ([#14746](https://github.com/tailwindlabs/tailwindcss/pull/14746))
```
Ah yep, I condensed it into a single entry, I like this, thanks!

RobinMalfait committed:
Add codemod for border style compatibility (#14746)

This PR adds a codemod that ensures that the border styles from Tailwind CSS v3 work as expected once your project is migrated to Tailwind CSS v4.

In Tailwind CSS v3, the default border color is `colors.gray.200` and in Tailwind CSS v4 the default border color is `currentColor`.

Similarly in Tailwind CSS v3, DOM elements such as `input`, `select`, and `textarea` have a border width of `0px`, in Tailwind CSS v4, we don't change the border width of these elements and keep them as `1px`.

If your project happens to already use the same value for the default border color (`currentColor`) as we use in Tailwind CSS v4, then nothing happens. But this is very unlikely, so we will make sure that we honor your `borderColor.DEFAULT` value.

If you didn't change the default values in your `tailwind.config.js`, then we will inject compatibility CSS using the default Tailwind CSS v3 values to ensure the default color and width are applied correctly.
